core/src/apps/nem/transfer/layout.py

```diff
@@ -22,15 +22,20 @@
 from ..layout import require_confirm_final, require_confirm_text
 from ..mosaic.helpers import get_mosaic_definition, is_nem_xem_mosaic
 
+if False:
+    from trezor.wire import Context
+    from ..mosaic.helpers import Mosaic
+
 
 async def ask_transfer(
-    ctx,
+    ctx: Context,
     common: NEMTransactionCommon,
     transfer: NEMTransfer,
     payload: bytes,
     encrypted: bool,
-):
+) -> None:
     if payload:
+        assert transfer.payload is not None
```

Comment:
Indeed a good question, I have no idea. Turns out `payload` and `transfer.payload` can be different when `transfer.public_key` is there - function `get_transfer_payload()`. 

in `serialize_transfer()`, which is the only other function using this `payload` variable, we also have access to `transfer`, but we _do_ use the `payload`:
https://github.com/trezor/trezor-firmware/blob/94cd8c25203985d6da236c4436be4ebbc24661f8/core/src/apps/nem/transfer/serialize.py#L44-L51

After changing it to use `payload`, the unittests `test_apps.nem.transfer.py` are still fine. Also, the `ask_transfer()` which uses this is just showing data to user, so I guess it is safe to change it.

Reply:
Changed in 329a038

Reply:
New plot-twist - I did not notice before, but it is breaking a device test `../tests/device_tests/test_msg_nem_signtx_transfers.py::TestMsgNEMSignTx::test_nem_signtx_encrypted_payload` on the `_require_confirm_payload` with `UnicodeError` which got me to find out how these two payloads even look, and below is the result. Therefore I see the reason to use `transfer.payload` in the layout. Will be reverting the change

```
payload b"J\r6\x1a!\x1c\xb3f\x0c\x87\xf5\x8f\xfb\x96\xe1\x8fb\x81\x9aqY\x91f3\xe4\xca>\xe8\xf3\x18\xba\xdc\xb6\t\x8e\x07\x9d\xf5I)X\xfaa\xc9\xd7\xb9\xbf\xf2\x85\xa2>\xe4\r\xf5'\x98^9Y&(\x90x\x87\x02\xa4g\x8f_$\x06\x86\x05\x1d\x19q\xc6c\xeb\xfd\xe6\x1f\xfd\x02 K)w\xb4\xd8\xfc3L\xd7*\x01"
transfer.payload b'this message should be encrypted'
```

Reply:
new question: why does the function accept `payload` in the first place?

i'm worried that the change to `get_transfer_payload` is changing the logic somehow

Reply:
Per comment on `get_transfer_payload`, we can remove the `payload` argument here and use `if transfer.payload` directly. It is guaranteed to be `bytes`, and if it is non-empty, it will be displayed for confirmation (and later serialized).

Reply:
Agreed, done in 8fcd268